### PR TITLE
luaPackages: add some lua packages

### PR DESCRIPTION
**Pull request by mrcjkb**
## Description of changes

@teto these are dependencies of rocks.nvim:

- luaPackages.toml (depends on sol2)
- luaPackages.toml-edit (depends on luaPackages.luarocks-build-rust-mlua)
- nui-nvim

toml-edit requires cargo to build, and toml tries to pull dependencies using git submodules, ~which is why I didn't put them in the generated-packages.nix~.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<details>
 <summary>Output of `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`</summary>

```console
21 packages added:
lua51Packages.luarocks-build-rust-mlua (init at 0.1.2-1) lua51Packages.nui-nvim (init at 0.2.0-1) lua51Packages.toml (init at 0.3.0-0) lua51Pa
ckages.toml-edit (init at 0.1.4-1) lua52Packages.luarocks-build-rust-mlua (init at 0.1.2-1) lua52Packages.nui-nvim (init at 0.2.0-1) lua52Pack
ages.toml (init at 0.3.0-0) lua52Packages.toml-edit (init at 0.1.4-1) lua53Packages.luarocks-build-rust-mlua (init at 0.1.2-1) lua53Packages.n
ui-nvim (init at 0.2.0-1) lua53Packages.toml (init at 0.3.0-0) lua53Packages.toml-edit (init at 0.1.4-1) lua54Packages.luarocks-build-rust-mlu
a (init at 0.1.2-1) lua54Packages.nui-nvim (init at 0.2.0-1) lua54Packages.toml (init at 0.3.0-0) lua54Packages.toml-edit (init at 0.1.4-1) lu
ajitPackages.luarocks-build-rust-mlua (init at 0.1.2-1) luajitPackages.nui-nvim (init at 0.2.0-1) luajitPackages.toml (init at 0.3.0-0) luajit
Packages.toml-edit (init at 0.1.4-1) sol2 (init at 3.3.1)

$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/mrcjk/.cache/nixpkgs-review
/rev-c701eeb3170af4d5ee64e9abdf8b5df9aa571d00/build.nix
error: builder for '/nix/store/39wlbyjcshl5fg2advj8xyw2y68yyjbi-review-shell.drv' failed with exit code 1;
       last 8 log lines:
       > unpacking sources
       > patching sources
       > Executing cargoSetupPostPatchHook
       > Validating consistency between /build/Cargo.lock and /Cargo.lock
       > /nix/store/q56n7lhjw724i7b33qaqra61p7m7c0cd-diffutils-3.10/bin/diff: /build/Cargo.lock: No such file or directory
       > /nix/store/q56n7lhjw724i7b33qaqra61p7m7c0cd-diffutils-3.10/bin/diff: /Cargo.lock: No such file or directory
       > ERROR: Missing Cargo.lock from src. Expected to find it at: /build/Cargo.lock
       > Hint: You can use the cargoPatches attribute to add a Cargo.lock manually to the build.
       For full logs, run 'nix log /nix/store/39wlbyjcshl5fg2advj8xyw2y68yyjbi-review-shell.drv'.
21 packages built:
lua51Packages.luarocks-build-rust-mlua lua51Packages.nui-nvim lua51Packages.toml lua51Packages.toml-edit lua52Packages.luarocks-build-rust-mlu
a lua52Packages.nui-nvim lua52Packages.toml lua52Packages.toml-edit lua53Packages.luarocks-build-rust-mlua lua53Packages.nui-nvim lu<details>a53Package
s.toml lua53Packages.toml-edit lua54Packages.luarocks-build-rust-mlua lua54Packages.nui-nvim lua54Packages.toml lua54Packages.toml-edit luajit
Packages.luarocks-build-rust-mlua luajitPackages.nui-nvim luajitPackages.toml luajitPackages.toml-edit sol2
```
</details>

I'm not sure what to do with the error message in `nixpkgs-review`.
All packages I added build successfully on my machine.

### CODE REVIEW

**Comment by mrcjkb** [comment]
It looks like there are other lua packages that depend on `luarocks-build-rust-mlua` (the `toml-edit` build dependency which I added, because it was missing) in generated-packages.nix. I suppose the reported build failure is one or more of those packages that nixpkgs-review is trying to build because I added it?

**Comment by mrcjkb** [comment]
Converting to draft, because I also need to add `toml`, which depends on `sol2`, which has invalid emoji characters in the test sources (╯°□°）╯︵ ┻━┻)

File: pkgs/development/libraries/sol2/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{ fetchFromGitHub
+, lib
+, stdenv
+, cmake
+, lua
+}:
+stdenv.mkDerivation rec{
```
Comment (teto) [inline pkgs/development/libraries/sol2/default.nix:7]: I like the `finalAttrs` convention used recenty that avoids the `rec` here

File: pkgs/development/lua-modules/toml-edit/default.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
+{ lib
```
Comment (teto) [inline pkgs/development/lua-modules/toml-edit/default.nix:1]: I would prefer if the overrides could be in pkgs/development/lua-modules/overrides.nix. This way the src keeps getting updated in generated.nix. Was there any difficulty doing so ?

File: pkgs/development/lua-modules/toml/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
+{ lib
```
Comment (teto) [inline pkgs/development/lua-modules/toml/default.nix:1]: same remark

**Review by teto — APPROVED** [review]

File: pkgs/development/libraries/sol2/default.nix (hunk 1 above)
Comment (mrcjkb) [reply pkgs/development/libraries/sol2/default.nix]: TIL

File: pkgs/development/lua-modules/toml-edit/default.nix (hunk 2 above)
Comment (mrcjkb) [reply pkgs/development/lua-modules/toml-edit/default.nix]: Yup, that works and is much cleaner. Thanks for the tip!

**Comment by teto** [comment]
Result of `nixpkgs-review pr 260759` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages built:</summary>
  <ul>
    <li>lua51Packages.luarocks-build-rust-mlua</li>
    <li>lua51Packages.nui-nvim</li>
    <li>lua51Packages.toml</li>
    <li>lua51Packages.toml-edit</li>
    <li>lua52Packages.luarocks-build-rust-mlua</li>
    <li>lua52Packages.nui-nvim</li>
    <li>lua52Packages.toml</li>
    <li>lua52Packages.toml-edit</li>
    <li>lua53Packages.luarocks-build-rust-mlua</li>
    <li>lua53Packages.nui-nvim</li>
    <li>lua53Packages.toml</li>
    <li>lua53Packages.toml-edit</li>
    <li>lua54Packages.luarocks-build-rust-mlua</li>
    <li>lua54Packages.nui-nvim</li>
    <li>lua54Packages.toml</li>
    <li>lua54Packages.toml-edit</li>
    <li>luajitPackages.luarocks-build-rust-mlua</li>
    <li>luajitPackages.nui-nvim</li>
    <li>luajitPackages.toml</li>
    <li>luajitPackages.toml-edit</li>
    <li>sol2</li>
  </ul>
</details>

**Pull request merged**